api/audit/models.py

```diff
@@ -99,9 +134,12 @@ def get_history_record_model_class(
         return getattr(module, class_name)
 
     @hook(BEFORE_CREATE)
-    def add_project(self):
-        if self.environment and self.project is None:
-            self.project = self.environment.project
+    def add_project_organisation(self):
+        """Ensure dependent fields are present"""
+        if self.project_id is None and self.environment:
+            self.project_id = self.environment.project_id
+        if self.organisation_id is None and self.project:
+            self.organisation_id = self.project.organisation_id
```

Comment:
The type checker is happier with this because it doesn't know that if `self.environment_id` is truthy then `self.environment` is not `None`.
Functionality-wise I'd check e.g. `self.environment_id` instead of `self.environment` in order to avoid hitting the database to load the `Environment` object, but the next line will access `self.environment` anyway, so I don't think this will affect functionality either way.